myst_nb/parser.py

```diff
@@ -21,6 +27,94 @@ class NotebookParser(MystParser):
     config_section_dependencies = ("parsers",)
 
     def parse(self, inputstring, document):
+
+        # de-serialize the notebook
+        ntbk = nbf.reads(inputstring, nbf.NO_CONVERT)
+
+        # This is a contaner for top level markdown tokens
+        # which we will add to as we walk the document
+        mkdown_tokens = []
+
+        # first we ensure that we are using a 'clean' global context
+        # for parsing, which is setup with the MyST parsing tokens
+        parse_context = ParseContext(
+            find_blocks=SphinxNBRenderer.default_block_tokens,
+            find_spans=SphinxNBRenderer.default_span_tokens,
+        )
+        # TODO hook in the document.reporter here,
+        # to e.g. report on duplicate link/footnote definitions
+        set_parse_context(parse_context)
+
+        for cell_index, nb_cell in enumerate(ntbk.cells):
+
+            # Skip empty cells
+            if len(nb_cell["source"]) == 0:
+                continue
+
+            # skip cells tagged for removal
+            tags = nb_cell.metadata.get("tags", [])
+            if "remove_cell" in tags:
+                continue
+
+            if nb_cell["cell_type"] == "markdown":
+                lines = nb_cell["source"]
+
+                # ensure all lines end with a new line
+                # TODO this is going to be changed in mistletoe, to just be
+                # lines = SourceLines(lines)
+                if isinstance(lines, str):
+                    lines = lines.splitlines(keepends=True)
+                lines = [
+                    line if line.endswith("\n") else "{}\n".format(line)
+                    for line in lines
+                ]
+
+                # parse the markdown;
+                # at this point span/inline level tokens are not yet processed, but
+                # link/footnote definitions are collected/stored in the global context
+                # TODO here it would be ideal to somehow include the cell index
+                # in the `position` attribute of each token
+                # and utilise in within the document.reporter
+                mkdown_tokens.extend(tokenize_block(lines))
```

Comment:
And the *definitions* are always block-level, yeah?

Reply:
Yes [LinkDefinition](https://mistletoe-ebp.readthedocs.io/en/latest/api/core_block_tokens.html#linkdefinition) and [Footnote](https://mistletoe-ebp.readthedocs.io/en/latest/api/extension_tokens.html#footnote) are block tokens